design/login.md

```diff
@@ -0,0 +1,102 @@
+# login
+
+The login package handles DC/OS login related operations.
+
+## Goals
+
+The goals of the login package are to :
+
+- Detect available login providers for a DC/OS cluster.
+- Log in to a DC/OS cluster using a given login provider.
+
+## Login providers detection
+
+To detect available login providers for a given DC/OS cluster, the CLI issues a GET request to the
+`/acs/api/v1/auth/providers` endpoint. This endpoint returns a list of providers in JSON. Each
+top-level key refers to the login provider ID, while its value is the login provider definition:
+
+``` json
+{
+  "dcos-users": {
+    "authentication-type": "dcos-uid-password",
+    "client-method": "dcos-usercredential-post-receive-authtoken",
+    "config": {
+      "start_flow_url": "/acs/api/v1/auth/login"
+    },
+    "description": "Default DC/OS authenticator"
+  }
+}
+```
+
+The above definition is a default DC/OS login provider, its ID is **dcos-users**.
+The ID has no special semantic apart from uniquely identifying a login provider.
+
+When the `/acs/api/v1/auth/providers` endpoint is not present, the available login provider is
+detected by making an unauthenticated `HEAD` request to a well-known protected DC/OS resource and
+reading the `WWW-Authenticate` header DC/OS returns in the 401 response:
+
+- `acsjwt` : the cluster only supports the `dcos-users` provider as defined above.
+- `oauthjwt` : the cluster only supports the OpenID implicit flow provider, as defined below:
+
+``` json
+{
+  "dcos-oidc-auth0": {
+    "authentication-type": "oidc-implicit-flow",
+    "client-method": "browser-prompt-authtoken",
+    "config": {
+      "start_flow_url": "/login?redirect_uri=urn:ietf:wg:oauth:2.0:oob"
+    },
+    "description": "Google, GitHub, or Microsoft"
+  }
+}
+```
+
+## Log in to a DC/OS cluster
+
+Log in to a DC/OS cluster is the action of presenting some credentials to the cluster in order to
+get an authorization token in return (also known as `ACS token`).
+
+The DC/OS CLI supports 6 login providers types (`authentication-type`):
+
+- **dcos-uid-password** : Log in using a standard DC/OS user account (username and password).
+- **dcos-uid-password-ldap** : Log in using an LDAP user account (username and password).
+- **dcos-uid-servicekey** : Log in using a DC/OS service user account (username and private key).
+- **saml-sp-initiated** : Log in using SAML 2.0.
+- **oidc-authorization-code-flow** : Log in using OpenID Connect authorization code flow.
+- **oidc-implicit-flow** : Log in using OpenID Connect implicit flow.
+
+A login happens through the `dcos auth login` command, which accepts the following flags:
+
+- **--provider** (string) : Specify the login provider ID to use.
+- **--username** (string) : Specify the username.
+- **--password** (string) : Specify password on the command line (insecure).
```

Comment:
`on the command line` seems unnecessary.

Reply:
`Specify password (insecure).` sounds unclear, in my mind here `on the command line` emphasizes on why it is `insecure`.

Reply:
I quickly looked around for examples from other command line utilities but it seems most will prompt for a password. I'd say wording it like `Specify password in plaintext (insecure)` might be better. Or maybe `Specify password inline in plaintext (insecure)` to be a little more explicit.

Reply:
Updated, I've also added back the `--password-file` option which I was planning to deprecate as suggested here : https://docs.google.com/document/d/17rR1BNEE9Ws27Faa5poPnpitqhT1ks843tCpSUD1Tu8/edit?disco=AAAAB3jFhV8

But the same argument could be said for the --password option, it most likely persists the password to the disk (in the shell history), so marking them as insecure sounds reasonable and avoids breaking users scripts in the future.